File: packages/cast/android/src/main/java/com/margelo/nitro/rnmediacast/HybridRnMediaCast.kt
// NOT a free choice: nitrogen hardcodes the implementation class's JNI
// descriptor as `Lcom/margelo/nitro/<androidNamespace>/<implementationClassName>;`
// (see nitrogen/generated/android/RnMediaCastOnLoad.cpp). Anywhere else and
// `NitroModules.createHybridObject('RnMediaCast')` throws
// ClassNotFoundException at runtime, with nothing failing at build time.
package com.margelo.nitro.rnmediacast

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.google.android.gms.cast.framework.CastContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.rnmediacast.CastController
import com.rnmediacast.CastEmitter
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicLong

/**
 * Android half of `@rn-media/cast`.
 *
 * Threading contract
 * ------------------
 * The Cast framework is main-thread-only, so every spec method trampolines
 * onto the main looper before touching it; the [CastController] is then
 * single-threaded by construction. Listener *registries* are concurrent maps
 * because `addXListener`/`removeXListener` arrive on the JS thread while
 * emissions come from main. Nitro schedules JS callback invocation onto the JS
 * thread itself (nitro.margelo.com/docs/types/callbacks).
 *
 * Availability contract
 * ---------------------
 * The framework is a dynamite module loaded from Google Play services at
 * runtime. On a GMS-less device `CastContext.getSharedInstance(Context,
 * Executor)`'s Task fails — that failure is caught and becomes the typed
 * `'unavailable'` state; it must never crash and never reject `initialize`.
 */
class HybridRnMediaCast : HybridRnMediaCastSpec(), CastEmitter {

  private val reactContext: ReactApplicationContext =
    NitroModules.applicationContext
      ?: throw IllegalStateException(
        "[cast] No ReactApplicationContext available. " +
          "Is react-native-nitro-modules installed correctly?"
      )

  /** Application context: this object outlives any Activity. */
  private val appContext: Context = reactContext.applicationContext

  private val handler = Handler(Looper.getMainLooper())
  private val mainExecutor = Executor { runnable -> handler.post(runnable) }

  /** Main-thread only. */
  private var controller: CastController? = null

  /**
   * Cache read by the synchronous [getCastState]. `UNAVAILABLE` until
   * [initialize] succeeds — before init, casting genuinely is not available.
   */
  @Volatile
  private var cachedState: CastConnectionState = CastConnectionState.UNAVAILABLE

  private val nextListenerId = AtomicLong(1)
  private val castStateListeners =
    ConcurrentHashMap<Long, (NativeCastStateEvent) -> Unit>()
  private val sessionListeners =
    ConcurrentHashMap<Long, (NativeCastSessionEvent) -> Unit>()
  private val devicesListeners =
    ConcurrentHashMap<Long, (NativeCastDevicesEvent) -> Unit>()
  private val mediaStatusListeners =
    ConcurrentHashMap<Long, (NativeCastMediaStatusEvent) -> Unit>()
  private val mediaErrorListeners =
    ConcurrentHashMap<Long, (NativeCastMediaErrorEvent) -> Unit>()
  private val queueChangedListeners = ConcurrentHashMap<Long, () -> Unit>()
  private val deviceVolumeListeners =
    ConcurrentHashMap<Long, (NativeDeviceVolumeEvent) -> Unit>()

  // -------------------------------------------------------------------------
  // Spec: context
  // -------------------------------------------------------------------------

  override fun initialize(
    receiverApplicationId: String?
  ): Promise<CastConnectionState> {
    val promise = Promise<CastConnectionState>()
    runOnMain {
      val existing = controller
      if (existing != null) {
        // Idempotent: a second initialize only re-applies the app id.
        receiverApplicationId?.let { existing.setReceiverApplicationId(it) }
        cachedState = existing.currentConnectionState()
        promise.resolve(cachedState)
        return@runOnMain
      }
      try {
        CastContext.getSharedInstance(appContext, mainExecutor)
          .addOnSuccessListener { castContext ->
            val created =
              CastController(
                appContext,
                castContext,
                this,
                handler
              ) { reactContext.currentActivity }
            controller = created
            receiverApplicationId?.let { created.setReceiverApplicationId(it) }
            cachedState = created.currentConnectionState()
            promise.resolve(cachedState)
          }
          .addOnFailureListener { error ->
            // No Play services / dynamite load failure / missing
            // OPTIONS_PROVIDER meta-data. A typed capability answer — logged,
            // never thrown, never swallowed.
            Log.w(TAG, "Cast framework unavailable", error)
            cachedState = CastConnectionState.UNAVAILABLE
            promise.resolve(CastConnectionState.UNAVAILABLE)
          }
      } catch (error: Throwable) {
        Log.w(TAG, "Cast framework unavailable (init threw)", error)
        cachedState = CastConnectionState.UNAVAILABLE
        promise.resolve(CastConnectionState.UNAVAILABLE)
      }
    }
    return promise
  }

  override fun getCastState(): CastConnectionState = cachedState

  override fun startDiscovery(): Promise<Unit> =
    withController { c, promise -> c.startDiscovery(promise) }

  override fun stopDiscovery(): Promise<Unit> =
    withController { c, promise -> c.stopDiscovery(promise) }

  override fun getDevices(): Promise<Array<CastDeviceInfo>> =
    withController { c, promise -> c.getDevices(promise) }

  override fun requestSession(deviceId: String): Promise<Unit> =
    withController { c, promise -> c.requestSession(deviceId, promise) }

  override fun showCastPicker(): Promise<Unit> =
    withController { c, promise -> c.showCastPicker(promise) }

  override fun endSession(stopReceiver: Boolean): Promise<Unit> =
    withController { c, promise -> c.endSession(stopReceiver, promise) }

  // -------------------------------------------------------------------------
  // Spec: media + volume
  // -------------------------------------------------------------------------

  override fun load(
    source: CastMediaSource,
    options: CastLoadOptions
  ): Promise<Unit> = withController { c, promise -> c.load(source, options, promise) }

  override fun play(): Promise<Unit> = withController { c, promise -> c.play(promise) }

  override fun pause(): Promise<Unit> = withController { c, promise -> c.pause(promise) }

  override fun stop(): Promise<Unit> = withController { c, promise -> c.stop(promise) }

  override fun seek(
    position: Double,
    resumeState: CastSeekResumeState
  ): Promise<Unit> = withController { c, promise -> c.seek(position, resumeState, promise) }

  override fun getApproximatePosition(): Promise<Double> =
    withController { c, promise -> c.getApproximatePosition(promise) }

  override fun setStreamVolume(volume: Double): Promise<Unit> =
    withController { c, promise -> c.setStreamVolume(volume, promise) }

  override fun setStreamMuted(muted: Boolean): Promise<Unit> =
    withController { c, promise -> c.setStreamMuted(muted, promise) }

  override fun setDeviceVolume(volume: Double): Promise<Unit> =
    withController { c, promise -> c.setDeviceVolume(volume, promise) }

  override fun setDeviceMuted(muted: Boolean): Promise<Unit> =
    withController { c, promise -> c.setDeviceMuted(muted, promise) }

  override fun getDeviceVolume(): Promise<NativeDeviceVolumeEvent> =
    withController { c, promise -> c.getDeviceVolume(promise) }

  // -------------------------------------------------------------------------
  // Spec: receiver queue
  // -------------------------------------------------------------------------

  override fun queueLoad(
    items: Array<CastQueueItemInput>,
    options: CastQueueLoadOptions
  ): Promise<Unit> = withController { c, promise -> c.queueLoad(items, options, promise) }

  override fun queueInsert(
    items: Array<CastQueueItemInput>,
    beforeItemId: Double?
  ): Promise<Unit> =
    withController { c, promise -> c.queueInsert(items, beforeItemId, promise) }

  override fun queueRemove(itemIds: DoubleArray): Promise<Unit> =
    withController { c, promise -> c.queueRemove(itemIds, promise) }

  override fun queueReorder(
    itemIds: DoubleArray,
    beforeItemId: Double?
  ): Promise<Unit> =
    withController { c, promise -> c.queueReorder(itemIds, beforeItemId, promise) }

  override fun queueJumpTo(itemId: Double, position: Double?): Promise<Unit> =
    withController { c, promise -> c.queueJumpTo(itemId, position, promise) }

  override fun queueSetRepeatMode(mode: CastRepeatMode): Promise<Unit> =
    withController { c, promise -> c.queueSetRepeatMode(mode, promise) }

  override fun getQueueItemIds(): Promise<DoubleArray> =
    withController { c, promise -> c.getQueueItemIds(promise) }

  override fun fetchQueueSlice(
    startIndex: Double,
    count: Double
  ): Promise<Array<CastQueueItemSnapshot>> =
    withController { c, promise -> c.fetchQueueSlice(startIndex, count, promise) }

  // -------------------------------------------------------------------------
  // Spec: listeners
  // -------------------------------------------------------------------------

  override fun addCastStateListener(
    listener: (event: NativeCastStateEvent) -> Unit
  ): Double = register(castStateListeners, listener)

  override fun removeCastStateListener(listenerId: Double) {
    castStateListeners.remove(listenerId.toLong())
  }

  override fun addSessionListener(
    listener: (event: NativeCastSessionEvent) -> Unit
  ): Double = register(sessionListeners, listener)

  override fun removeSessionListener(listenerId: Double) {
    sessionListeners.remove(listenerId.toLong())
  }

  override fun addDevicesListener(
    listener: (event: NativeCastDevicesEvent) -> Unit
  ): Double = register(devicesListeners, listener)

  override fun removeDevicesListener(listenerId: Double) {
    devicesListeners.remove(listenerId.toLong())
  }

  override fun addMediaStatusListener(
    listener: (event: NativeCastMediaStatusEvent) -> Unit
  ): Double = register(mediaStatusListeners, listener)

  override fun removeMediaStatusListener(listenerId: Double) {
    mediaStatusListeners.remove(listenerId.toLong())
  }

  override fun addMediaErrorListener(
    listener: (event: NativeCastMediaErrorEvent) -> Unit
  ): Double = register(mediaErrorListeners, listener)

  override fun removeMediaErrorListener(listenerId: Double) {
    mediaErrorListeners.remove(listenerId.toLong())
  }

  override fun addQueueChangedListener(listener: () -> Unit): Double =
    register(queueChangedListeners, listener)

  override fun removeQueueChangedListener(listenerId: Double) {
    queueChangedListeners.remove(listenerId.toLong())
  }

  override fun addDeviceVolumeListener(
    listener: (event: NativeDeviceVolumeEvent) -> Unit
  ): Double = register(deviceVolumeListeners, listener)

  override fun removeDeviceVolumeListener(listenerId: Double) {
    deviceVolumeListeners.remove(listenerId.toLong())
  }

  // -------------------------------------------------------------------------
  // CastEmitter (called from main by the controller)
  // -------------------------------------------------------------------------

  override fun onCastState(event: NativeCastStateEvent) {
    cachedState = event.state
    castStateListeners.values.forEach { it(event) }
  }

  override fun onSession(event: NativeCastSessionEvent) {
    sessionListeners.values.forEach { it(event) }
  }

  override fun onDevices(event: NativeCastDevicesEvent) {
    devicesListeners.values.forEach { it(event) }
  }

  override fun onMediaStatus(event: NativeCastMediaStatusEvent) {
    mediaStatusListeners.values.forEach { it(event) }
  }

  override fun onMediaError(event: NativeCastMediaErrorEvent) {
    mediaErrorListeners.values.forEach { it(event) }
  }

  override fun onQueueChanged() {
    queueChangedListeners.values.forEach { it() }
  }

  override fun onDeviceVolume(event: NativeDeviceVolumeEvent) {
    deviceVolumeListeners.values.forEach { it(event) }
  }

  // -------------------------------------------------------------------------
  // Internals
  // -------------------------------------------------------------------------

  private fun <T : Any> register(
    into: ConcurrentHashMap<Long, T>,
    listener: T
  ): Double {
    val id = nextListenerId.getAndIncrement()
    into[id] = listener
    return id.toDouble()
  }

  private fun runOnMain(block: () -> Unit) {
    if (Looper.myLooper() == Looper.getMainLooper()) block() else handler.post(block)
  }

  /**
   * Trampoline a controller call onto the main thread; before a successful
   * `initialize` every call is a typed `[unavailable]` rejection.
   */
  private fun <T> withController(
    block: (CastController, Promise<T>) -> Unit
  ): Promise<T> {
    val promise = Promise<T>()
    runOnMain {
      val c = controller
      if (c == null) {
        promise.reject(
          IllegalStateException(
            "[unavailable] Cast framework is not initialized — call initialize() " +
              "first (it resolves 'unavailable' on devices without Google Play services)."
          )
        )
        return@runOnMain
      }
      try {
        block(c, promise)
      } catch (error: Throwable) {
        promise.reject(
          IllegalStateException("[native] ${error.message ?: error.javaClass.simpleName}", error)
        )
      }
    }
    return promise
  }

  private companion object {
    const val TAG = "RnMediaCast"
  }
}

File: packages/cast/ios/HybridRnMediaCast.swift
//
//  HybridRnMediaCast.swift
//  RnMediaCast
//
//  The iOS half of `@rn-media/cast`.
//
//  Threading contract
//  ------------------
//  GoogleCast is main-thread-only, so every spec method trampolines onto the
//  main queue before touching it; the `CastCoordinator` is then
//  single-threaded by construction. Listener registries are lock-guarded
//  because `add*/remove*Listener` arrive on the JS thread while emissions
//  come from main. Nitro schedules JS callback invocation onto the JS thread
//  itself (nitro.margelo.com/docs/types/callbacks).
//
//  Init-timing ceiling (documented in the spec): Google's guidance is to call
//  `GCKCastContext.setSharedInstanceWith` in
//  `application(_:didFinishLaunchingWithOptions:)` so the SDK can resume a
//  session the app was killed during. JS-driven `initialize` runs later than
//  that, so post-process-death session resumption may be missed until the
//  first initialize of a launch.
//

import Foundation
import GoogleCast
import NitroModules

final class HybridRnMediaCast: HybridRnMediaCastSpec, CastEmitter {
  /// Main-thread only.
  private var coordinator: CastCoordinator?

  /// The receiver application id the shared `GCKCastContext` was created with.
  /// Main-thread only. Kept so a later, differing `initialize` can say out
  /// loud that iOS cannot honour it (see `initialize`).
  private var activeReceiverApplicationId: String?

  /// Cache read by the synchronous `getCastState`. `.unavailable` until
  /// `initialize` succeeds — before init, casting genuinely is not available.
  private var cachedState: CastConnectionState = .unavailable

  private let lock = NSLock()
  private var nextListenerId: Double = 1
  private var castStateListeners: [Double: (NativeCastStateEvent) -> Void] = [:]
  private var sessionListeners: [Double: (NativeCastSessionEvent) -> Void] = [:]
  private var devicesListeners: [Double: (NativeCastDevicesEvent) -> Void] = [:]
  private var mediaStatusListeners: [Double: (NativeCastMediaStatusEvent) -> Void] = [:]
  private var mediaErrorListeners: [Double: (NativeCastMediaErrorEvent) -> Void] = [:]
  private var queueChangedListeners: [Double: () -> Void] = [:]
  private var deviceVolumeListeners: [Double: (NativeDeviceVolumeEvent) -> Void] = [:]

  // MARK: - Spec: context

  func initialize(receiverApplicationId: String?) throws -> Promise<CastConnectionState> {
    let promise = Promise<CastConnectionState>()
    runOnMain { [self] in
      if let existing = coordinator {
        // Idempotent. CEILING: the GCK shared instance cannot change its
        // receiver app id after creation — GCKCastContext exposes only
        // `+setSharedInstanceWithOptions:` / `+isSharedInstanceInitialized`
        // (GCKCastContext.h, 4.8.6), there is no `setReceiverApplicationId`
        // (Android has one) and no way to swap `GCKDiscoveryCriteria` on the
        // live `GCKDiscoveryManager`. Ignoring a differing id in silence is
        // exactly the failure mode this package refuses, so say so.
        if let requested = receiverApplicationId, requested != activeReceiverApplicationId {
          let active = activeReceiverApplicationId ?? kGCKDefaultMediaReceiverApplicationID
          NSLog(
            "%@",
            "[RnMediaCast] initialize(receiverApplicationId: \"\(requested)\") was " +
              "ignored: iOS fixes the receiver application id at the first initialize " +
              "(currently \"\(active)\") and the Cast SDK offers no way to change it. " +
              "Pass the id on the first call, or set it via the Expo plugin. Android " +
              "honours a later change."
          )
        }
        let state = existing.currentConnectionState()
        withLock { cachedState = state }
        existing.emitCurrentState()
        promise.resolve(withResult: state)
        return
      }
      // `isSharedInstanceInitialized` is a class METHOD in the SDK
      // (`+ (BOOL)isSharedInstanceInitialized;`, GCKCastContext.h:92), not a
      // property — it must be called, not read.
      if !GCKCastContext.isSharedInstanceInitialized() {
        let appId = receiverApplicationId ?? kGCKDefaultMediaReceiverApplicationID
        let criteria = GCKDiscoveryCriteria(applicationID: appId)
        let options = GCKCastOptions(discoveryCriteria: criteria)
        // Failure-mode rule from the design doc (§5 backgrounding): an audio
        // sender must keep its session while backgrounded; the OS-side limits
        // are documented in the README.
        options.suspendSessionsWhenBackgrounded = false
        GCKCastContext.setSharedInstanceWith(options)
      }
      guard GCKCastContext.isSharedInstanceInitialized() else {
        withLock { cachedState = .unavailable }
        promise.resolve(withResult: .unavailable)
        return
      }
      let created = CastCoordinator(context: GCKCastContext.sharedInstance(), emitter: self)
      coordinator = created
      activeReceiverApplicationId =
        receiverApplicationId ?? kGCKDefaultMediaReceiverApplicationID
      let state = created.currentConnectionState()
      withLock { cachedState = state }
      // The SDK may have resumed a session before JS got here, and nothing
      // replays that to a late listener. See CastCoordinator.emitCurrentState.
      created.emitCurrentState()
      promise.resolve(withResult: state)
    }
    return promise
  }

  func getCastState() throws -> CastConnectionState {
    lock.lock()
    defer { lock.unlock() }
    return cachedState
  }

  func startDiscovery() throws -> Promise<Void> {
    withCoordinator { $0.startDiscovery($1) }
  }

  func stopDiscovery() throws -> Promise<Void> {
    withCoordinator { $0.stopDiscovery($1) }
  }

  func getDevices() throws -> Promise<[CastDeviceInfo]> {
    withCoordinator { $0.getDevices($1) }
  }

  func requestSession(deviceId: String) throws -> Promise<Void> {
    withCoordinator { $0.requestSession(deviceId: deviceId, promise: $1) }
  }

  func showCastPicker() throws -> Promise<Void> {
    withCoordinator { $0.showCastPicker($1) }
  }

  func endSession(stopReceiver: Bool) throws -> Promise<Void> {
    withCoordinator { $0.endSession(stopReceiver: stopReceiver, promise: $1) }
  }

  // MARK: - Spec: media + volume

  func load(source: CastMediaSource, options: CastLoadOptions) throws -> Promise<Void> {
    withCoordinator { $0.load(source: source, options: options, promise: $1) }
  }

  func play() throws -> Promise<Void> { withCoordinator { $0.play($1) } }

  func pause() throws -> Promise<Void> { withCoordinator { $0.pause($1) } }

  func stop() throws -> Promise<Void> { withCoordinator { $0.stop($1) } }

  func seek(position: Double, resumeState: CastSeekResumeState) throws -> Promise<Void> {
    withCoordinator { $0.seek(position: position, resumeState: resumeState, promise: $1) }
  }

  func getApproximatePosition() throws -> Promise<Double> {
    withCoordinator { $0.getApproximatePosition($1) }
  }

  func setStreamVolume(volume: Double) throws -> Promise<Void> {
    withCoordinator { $0.setStreamVolume(volume, promise: $1) }
  }

  func setStreamMuted(muted: Bool) throws -> Promise<Void> {
    withCoordinator { $0.setStreamMuted(muted, promise: $1) }
  }

  func setDeviceVolume(volume: Double) throws -> Promise<Void> {
    withCoordinator { $0.setDeviceVolume(volume, promise: $1) }
  }

  func setDeviceMuted(muted: Bool) throws -> Promise<Void> {
    withCoordinator { $0.setDeviceMuted(muted, promise: $1) }
  }

  func getDeviceVolume() throws -> Promise<NativeDeviceVolumeEvent> {
    withCoordinator { $0.getDeviceVolume($1) }
  }

  // MARK: - Spec: receiver queue

  func queueLoad(items: [CastQueueItemInput], options: CastQueueLoadOptions) throws -> Promise<Void> {
    withCoordinator { $0.queueLoad(items: items, options: options, promise: $1) }
  }

  func queueInsert(items: [CastQueueItemInput], beforeItemId: Double?) throws -> Promise<Void> {
    withCoordinator { $0.queueInsert(items: items, beforeItemId: beforeItemId, promise: $1) }
  }

  func queueRemove(itemIds: [Double]) throws -> Promise<Void> {
    withCoordinator { $0.queueRemove(itemIds: itemIds, promise: $1) }
  }

  func queueReorder(itemIds: [Double], beforeItemId: Double?) throws -> Promise<Void> {
    withCoordinator { $0.queueReorder(itemIds: itemIds, beforeItemId: beforeItemId, promise: $1) }
  }

  func queueJumpTo(itemId: Double, position: Double?) throws -> Promise<Void> {
    withCoordinator { $0.queueJumpTo(itemId: itemId, position: position, promise: $1) }
  }

  func queueSetRepeatMode(mode: CastRepeatMode) throws -> Promise<Void> {
    withCoordinator { $0.queueSetRepeatMode(mode, promise: $1) }
  }

  func getQueueItemIds() throws -> Promise<[Double]> {
    withCoordinator { $0.getQueueItemIds($1) }
  }

  func fetchQueueSlice(startIndex: Double, count: Double) throws -> Promise<[CastQueueItemSnapshot]> {
    withCoordinator { $0.fetchQueueSlice(startIndex: startIndex, count: count, promise: $1) }
  }

  // MARK: - Spec: listeners

  func addCastStateListener(listener: @escaping (_ event: NativeCastStateEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      castStateListeners[id] = listener
      return id
    }
  }

  func removeCastStateListener(listenerId: Double) throws {
    withLock { _ = castStateListeners.removeValue(forKey: listenerId) }
  }

  func addSessionListener(listener: @escaping (_ event: NativeCastSessionEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      sessionListeners[id] = listener
      return id
    }
  }

  func removeSessionListener(listenerId: Double) throws {
    withLock { _ = sessionListeners.removeValue(forKey: listenerId) }
  }

  func addDevicesListener(listener: @escaping (_ event: NativeCastDevicesEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      devicesListeners[id] = listener
      return id
    }
  }

  func removeDevicesListener(listenerId: Double) throws {
    withLock { _ = devicesListeners.removeValue(forKey: listenerId) }
  }

  func addMediaStatusListener(listener: @escaping (_ event: NativeCastMediaStatusEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      mediaStatusListeners[id] = listener
      return id
    }
  }

  func removeMediaStatusListener(listenerId: Double) throws {
    withLock { _ = mediaStatusListeners.removeValue(forKey: listenerId) }
  }

  func addMediaErrorListener(listener: @escaping (_ event: NativeCastMediaErrorEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      mediaErrorListeners[id] = listener
      return id
    }
  }

  func removeMediaErrorListener(listenerId: Double) throws {
    withLock { _ = mediaErrorListeners.removeValue(forKey: listenerId) }
  }

  func addQueueChangedListener(listener: @escaping () -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      queueChangedListeners[id] = listener
      return id
    }
  }

  func removeQueueChangedListener(listenerId: Double) throws {
    withLock { _ = queueChangedListeners.removeValue(forKey: listenerId) }
  }

  func addDeviceVolumeListener(listener: @escaping (_ event: NativeDeviceVolumeEvent) -> Void) throws -> Double {
    withLock {
      let id = takeListenerId()
      deviceVolumeListeners[id] = listener
      return id
    }
  }

  func removeDeviceVolumeListener(listenerId: Double) throws {
    withLock { _ = deviceVolumeListeners.removeValue(forKey: listenerId) }
  }

  // MARK: - CastEmitter (called on main by the coordinator)

  func onCastState(_ event: NativeCastStateEvent) {
    let listeners = withLock { () -> [(NativeCastStateEvent) -> Void] in
      cachedState = event.state
      return Array(castStateListeners.values)
    }
    for listener in listeners { listener(event) }
  }

  func onSession(_ event: NativeCastSessionEvent) {
    for listener in withLock({ Array(sessionListeners.values) }) { listener(event) }
  }

  func onDevices(_ event: NativeCastDevicesEvent) {
    for listener in withLock({ Array(devicesListeners.values) }) { listener(event) }
  }

  func onMediaStatus(_ event: NativeCastMediaStatusEvent) {
    for listener in withLock({ Array(mediaStatusListeners.values) }) { listener(event) }
  }

  func onMediaError(_ event: NativeCastMediaErrorEvent) {
    for listener in withLock({ Array(mediaErrorListeners.values) }) { listener(event) }
  }

  func onQueueChanged() {
    for listener in withLock({ Array(queueChangedListeners.values) }) { listener() }
  }

  func onDeviceVolume(_ event: NativeDeviceVolumeEvent) {
    for listener in withLock({ Array(deviceVolumeListeners.values) }) { listener(event) }
  }

  // MARK: - Helpers

  private func runOnMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.async(execute: block)
    }
  }

  /// Trampoline a coordinator call onto the main thread; before a successful
  /// `initialize` every call is a typed `[unavailable]` rejection.
  private func withCoordinator<T>(
    _ block: @escaping (CastCoordinator, Promise<T>) -> Void
  ) -> Promise<T> {
    let promise = Promise<T>()
    runOnMain { [self] in
      guard let coordinator = self.coordinator else {
        promise.reject(withError: CastBridgeError.message(
          "[unavailable] Cast framework is not initialized — call initialize() first."))
        return
      }
      block(coordinator, promise)
    }
    return promise
  }

  private func withLock<T>(_ body: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return body()
  }

  /// Caller must hold `lock`.
  private func takeListenerId() -> Double {
    let id = nextListenerId
    nextListenerId += 1
    return id
  }
}
